compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs

```diff
@@ -2263,15 +2263,117 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
                 )
                 .unwrap_or_else(|guar| Const::new_error(tcx, guar))
             }
-            hir::ConstArgKind::Struct(..) => {
-                span_bug!(const_arg.span(), "lowering `{:?}` is not yet implemented", const_arg)
+            hir::ConstArgKind::Struct(qpath, inits) => {
+                self.lower_struct_expr_const_arg(hir_id, qpath, inits, const_arg.span())
             }
             hir::ConstArgKind::Anon(anon) => self.lower_anon_const(anon),
             hir::ConstArgKind::Infer(span, ()) => self.ct_infer(None, span),
             hir::ConstArgKind::Error(_, e) => ty::Const::new_error(tcx, e),
         }
     }
 
+    fn lower_struct_expr_const_arg(
+        &self,
+        hir_id: HirId,
+        qpath: hir::QPath<'tcx>,
+        inits: &'tcx [&'tcx hir::ConstArgExprField<'tcx>],
+        span: Span,
+    ) -> Const<'tcx> {
+        let tcx = self.tcx();
+
+        let non_adt_or_variant_res = || {
```

Comment:
add `FIXME` that most of this function should be reused with normal struct expr hir lowering